File: creme/assistants/models/action.py
# -*- coding: utf-8 -*-

################################################################################
#    Creme is a free/open-source Customer Relationship Management software
#    Copyright (C) 2009-2018  Hybird
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
################################################################################

import warnings

# from django.contrib.contenttypes.fields import GenericForeignKey
# from django.contrib.contenttypes.models import ContentType
from django.db import models
from django.urls import reverse
from django.utils.translation import ugettext_lazy as _

from creme.creme_core import models as creme_models
from creme.creme_core.models import fields as creme_fields


class ActionManager(models.Manager):
    def filter_by_user(self, user):
        return self.filter(user__in=[user] + user.teams)


class Action(creme_models.CremeModel):
    title             = models.CharField(_('Title'), max_length=200)
    is_ok             = models.BooleanField(_('Expected reaction has been done'), editable=False, default=False)
    description       = models.TextField(_('Source action'), blank=True)
    creation_date     = creme_fields.CreationDateTimeField(_('Creation date'), editable=False)
    expected_reaction = models.TextField(_('Target action'), blank=True)
    deadline          = models.DateTimeField(_('Deadline'))
    validation_date   = models.DateTimeField(_('Validation date'), blank=True, null=True, editable=False)
    user              = creme_fields.CremeUserForeignKey(verbose_name=_('Owner user'))

    # entity_content_type = models.ForeignKey(ContentType, related_name='action_entity_set', editable=False, on_delete=models.CASCADE)
    # entity_id           = models.PositiveIntegerField(editable=False).set_tags(viewable=False)
    # creme_entity        = GenericForeignKey(ct_field="entity_content_type", fk_field='entity_id')
    entity_content_type = creme_fields.EntityCTypeForeignKey(related_name='+', editable=False)
    entity              = models.ForeignKey(creme_models.CremeEntity, related_name='assistants_actions',
                                            editable=False, on_delete=models.CASCADE,
                                           ).set_tags(viewable=False)
    creme_entity        = creme_fields.RealEntityForeignKey(ct_field='entity_content_type', fk_field='entity')

    objects = ActionManager()

    creation_label = _('Create an action')
    save_label     = _('Save the action')

    class Meta:
        app_label = 'assistants'
        verbose_name = _('Action')
        verbose_name_plural = _('Actions')

    def __str__(self):
        return self.title

    def get_edit_absolute_url(self):
        return reverse('assistants__edit_action', args=(self.id,))

    @staticmethod
    def get_actions_it(entity, today):
        warnings.warn('Action.get_actions_it() is deprecated.', DeprecationWarning)
        return Action.objects.filter(entity_id=entity.id, is_ok=False, deadline__gt=today) \
                             .select_related('user')

    @staticmethod
    def get_actions_nit(entity, today):
        warnings.warn('Action.get_actions_nit() is deprecated.', DeprecationWarning)
        return Action.objects.filter(entity_id=entity.id, is_ok=False, deadline__lte=today) \
                             .select_related('user')

    @staticmethod
    def get_actions_it_for_home(user, today):
        warnings.warn('Action.get_actions_it_for_home() is deprecated.', DeprecationWarning)
        return Action.objects.filter(is_ok=False,
                                     deadline__gt=today,
                                     user__in=[user] + user.teams,
                                     # entity__is_deleted=False,
                                    ) \
                             .select_related('user')

    @staticmethod
    def get_actions_nit_for_home(user, today):
        warnings.warn('Action.get_actions_nit_for_home() is deprecated.', DeprecationWarning)
        return Action.objects.filter(is_ok=False,
                                     deadline__lte=today,
                                     user__in=[user] + user.teams,
                                     # entity__is_deleted=False,
                                    ) \
                             .select_related('user')

    @staticmethod
    def get_actions_it_for_ctypes(ct_ids, user, today):
        warnings.warn('Action.get_actions_it_for_ctypes() is deprecated.', DeprecationWarning)
        return Action.objects.filter(entity_content_type__in=ct_ids,
                                     user__in=[user] + user.teams,
                                     is_ok=False,
                                     deadline__gt=today,
                                    ) \
                             .select_related('user')

    @staticmethod
    def get_actions_nit_for_ctypes(ct_ids, user, today):
        warnings.warn('Action.get_actions_nit_for_ctypes() is deprecated.', DeprecationWarning)
        return Action.objects.filter(entity_content_type__in=ct_ids,
                                     user__in=[user] + user.teams,
                                     is_ok=False,
                                     deadline__lte=today
                                    ) \
                             .select_related('user')

    def get_related_entity(self):  # For generic views
        return self.creme_entity

File: creme/creme_core/static/creme_core/js/widgets/component/action.js
 /*******************************************************************************
 * Creme is a free/open-source Customer Relationship Management software
 * Copyright (C) 2009-2017 Hybird
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

(function($) {
"use strict";

var _ActionStatus = {
    DONE: 'done',
    RUNNING: 'run',
    FAIL: 'fail',
    CANCEL: 'cancel'
};

creme.component.ActionStatus = $.extend({}, _ActionStatus);

// TODO: 'creme.component' =>  'creme.action' (already defined in action-link.js)
creme.component.Action = creme.component.Component.sub({
    _init_: function(action, options) {
        this._events = new creme.component.EventHandler();
        this._options = options || {};
        this._status = _ActionStatus.DONE;
        this._action = Object.isFunc(action) ? action : this.done;
    },

    start: function() {
        if (this.isRunning() === true) {
            return this;
        }

        try {
            this._events.trigger('start', Array.copy(arguments), this);
            this._status = _ActionStatus.RUNNING;
            this._action.apply(this, Array.copy(arguments));
        } catch (e) {
            console.error(e);
            this.fail(e);
        }

        return this;
    },

    done: function() {
        if (this.isRunning() === false) {
            return this;
        }

        this._status = _ActionStatus.DONE;
        this._events.trigger('done', Array.copy(arguments), this);
        return this;
    },

    fail: function() {
        if (this.isRunning() === false) {
            return this;
        }

        this._status = _ActionStatus.FAIL;
        this._events.trigger('fail', Array.copy(arguments), this);
        return this;
    },

    cancel: function() {
        if (this.isRunning() === false) {
            return this;
        }

        this._status = _ActionStatus.CANCEL;
        this._events.trigger('cancel', Array.copy(arguments), this);
        return this;
    },

    trigger: function(event) {
        this._events.trigger(event, Array.copy(arguments).slice(1), this);
        return this;
    },

    action: function(data) {
        var self = this;
        var action = (data === undefined || Object.isFunc(data)) ? data : function() { self.done(data); };
        return Object.property(this, '_action', action);
    },

    one: function(key, listener, decorator) {
        this._events.one(key, listener, decorator);
        return this;
    },

    on: function(key, listener, decorator) {
        return this.bind(key, listener, decorator);
    },

    off: function(key, listener) {
        return this.unbind(key, listener);
    },

    bind: function(key, listener, decorator) {
        this._events.bind(key, listener, decorator);
        return this;
    },

    unbind: function(key, listener) {
        this._events.unbind(key, listener);
        return this;
    },

    onStart: function(start) {
        return this.bind('start', start);
    },

    onComplete: function(complete) {
        return this.bind('done fail cancel', complete);
    },

    onDone: function(success) {
        return this.bind('done', success);
    },

    onCancel: function(canceled) {
        return this.bind('cancel', canceled);
    },

    onFail: function(error) {
        return this.bind('fail', error);
    },

    isRunning: function() {
        return this._status === _ActionStatus.RUNNING;
    },

    isStatusDone: function() {
        return this._status === _ActionStatus.DONE;
    },

    isStatusFail: function() {
        return this._status === _ActionStatus.FAIL;
    },

    isStatusCancel: function() {
        return this._status === _ActionStatus.CANCEL;
    },

    status: function() {
        return this._status;
    },

    options: function(options) {
        return Object.property(this, '_options', options);
    },

    listen: function(source) {
        return this.after(source);
    },

    after: function(source) {
        var self = this;

        source.onDone(function() {
            self.start.apply(self, [source].concat(Array.copy(arguments).slice(1)));
        });

        source.onFail(function(event) {
            self._status = _ActionStatus.FAIL;
            self._events.trigger('fail', [source].concat(Array.copy(arguments).slice(1)), self);
        });

        source.onCancel(function(event) {
            self._status = _ActionStatus.CANCEL;
            self._events.trigger('cancel', [source].concat(Array.copy(arguments).slice(1)), self);
        });

        return this;
    },

    before: function(target) {
        target.after(this);
        return this;
    }
});

creme.component.TimeoutAction = creme.component.Action.sub({
    _init_: function(options) {
        this._super_(creme.component.Action, '_init_', this._startTimeout, options);
    },

    _startTimeout: function(options) {
        options = $.extend({}, this.options(), options);

        var self = this;
        var delay = options.delay || 0;

        if (delay > 0) {
            window.setTimeout(function() {
                if (self.isRunning()) {
                    self.done(options);
                }
            }, delay);
        } else {
            self.done(options);
        }
    }
});

}(jQuery));
